utils/id.go

```diff
@@ -66,17 +73,91 @@ func LocalNodeID() (string, error) {
 	return fmt.Sprintf("%s%s", NodePrefix, HashedID(hostname)[:8]), nil
 }
 
-var b62Index = newB62Index()
-var b62Chars = []byte(shortuuid.DefaultAlphabet)
+var b57Index = newB57Index()
+var b57Chars = []byte(shortuuid.DefaultAlphabet)
 
-func newB62Index() [256]byte {
+func newB57Index() [256]byte {
 	var index [256]byte
-	for i := 0; i < len(b62Chars); i++ {
-		index[b62Chars[i]] = byte(i)
+	for i := 0; i < len(b57Chars); i++ {
+		index[b57Chars[i]] = byte(i)
 	}
 	return index
 }
 
+type guidGenerator struct {
+	pool sync.Pool
+	mu   sync.Mutex
+	buf  []byte
+	pos  int
+}
+
+func newGuidGenerator(bufSize, scratchSize int) *guidGenerator {
+	return &guidGenerator{
+		pool: sync.Pool{
+			New: func() any {
+				b := make([]byte, scratchSize)
+				return &b
+			},
+		},
+		buf: make([]byte, bufSize),
+		pos: bufSize,
+	}
+}
+
+func (g *guidGenerator) refillBuf() {
+	for {
+		_, err := rand.Read(g.buf)
+		if err == nil {
+			return
+		}
+
+		logger.Errorw("unable to refill guid buffer", err)
+		time.Sleep(time.Millisecond)
+	}
+}
+
+func (g *guidGenerator) uint32() uint32 {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	if g.pos == len(g.buf) {
+		g.refillBuf()
+		g.pos = 0
+	}
+
+	n := binary.BigEndian.Uint32(g.buf[g.pos:])
+	g.pos += 4
+
+	return n
+}
+
+func (g *guidGenerator) readIDChars(b []byte) {
+	var n int
+	for {
+		r := g.uint32()
+		for i := 0; i < 5; i++ {
+			if int(r&0x3f) < len(b57Chars) {
+				b[n] = b57Chars[r&0x3f]
+				n++
+				if n == len(b) {
+					return
+				}
+			}
+			r >>= 6
+		}
+	}
+}
+
+func (g *guidGenerator) NewGuid(prefix string) string {
+	b := g.pool.Get().(*[]byte)
+	defer g.pool.Put(b)
```

Comment:
not following here.. if we've put it back after returning it cast as a string, would that cause it to be overridden by the next `NewGuid` call?

Reply:
strings are immutable and bytes are not so to avoid confusion when you cast from one to the other go always allocates. to get a string that references the memory from a byte you have to do ```
s := unsafe.String(unsafe.SliceData(b), len(b))
```

Reply:
got it. that makes sense! I had incorrectly assumed it was a straight cast.